Compiler/src/typeinfer.jl

```diff
@@ -115,7 +115,7 @@ function finish!(interp::AbstractInterpreter, caller::InferenceState, validation
             store_backedges(ci, edges)
         end
         inferred_result = nothing
-        uncompressed = inferred_result
+        uncompressed = result.src
         const_flag = is_result_constabi_eligible(result)
         discard_src = caller.cache_mode === CACHE_MODE_NULL || const_flag
         if !discard_src
```

Comment:
It looks like the implementation of every after here assumes that `transform_result_for_cache` returns a CodeInfo for correctness, and needs to be re-written

Reply:
It looks like debuginfo may still computed incorrectly on this part of the PR

Reply:
@serenity4 take a look?

Although independently it does now feel a bit weird from a dataflow perspective that the debuginfo that goes with the optimized source takes a different path.

Reply:
It would make more sense to me to take `di = src.debuginfo` (where `src::CodeInfo` is extracted from `result.src::Union{CodeInfo, OptimizationState}`) and override it with what we can extract from `inferred_result` (essentially doing the same logic as for `result.src`).

Reply:
Perhaps something like that: https://github.com/serenity4/julia/commit/318b9a34e8c24c0faab7e1456d8f967e97bf46e0

Reply:
I've rebased this PR and cherry-picked your commit.